macOS/DuckDuckGo/Autoconsent/AutoconsentUserScript.swift

```diff
@@ -543,4 +575,61 @@ extension AutoconsentUserScript {
         // fire daily pixel if needed
         PixelKit.fire(pixel, frequency: .daily)
     }
+
+    // MARK: - Reload Loop Detection
+
+    /// Detects a reload loop
+    /// - Parameters:
+    ///   - cmpName: The name of the CMP that was detected
+    private func detectReloadLoop(cmpName: String) {
+        // Reload loop is when we catch the same CMP from the same top URL without a navigation in between.
+        // At this point we know that the top URL hasn't changed (that's tracked in handleInit), so we can just check the CMP name.
+        if !reloadLoopDetected && cmpName == lastHandledCMPName {
+            // Same CMP detected on same URL after it was already handled - reload loop detected
+            Logger.autoconsent.debug("Reload loop detected: CMP \(cmpName) on \(String(describing: self.topUrl))")
+            reloadLoopDetected = true
+            firePixel(pixel: .errorReloadLoop)
+        }
+    }
+
+    /// Stores the URL and CMP name after a popup was successfully handled
+    /// - Parameters:
+    ///   - cmpName: The name of the CMP that was handled
+    ///   - isCosmetic: Whether this was a cosmetic rule (cosmetic rules don't trigger reload loops)
+    private func rememberLastHandledCMP(cmpName: String, isCosmetic: Bool) {
+        if isCosmetic {
+            // Cosmetic rules can trigger on every page load and never cause reload loops
+            Logger.autoconsent.debug("Cosmetic rule handled, not storing for reload loop detection")
+            clearReloadLoopState()
+            return
+        }
+
+        if lastHandledCMPName != cmpName {
+            // The last handled CMP is different from the current one, so we need to clear the reload loop state
+            Logger.autoconsent.debug("Last handled CMP is changed from \(self.lastHandledCMPName) to \(cmpName), clearing reload loop state")
+            clearReloadLoopState()
+        }
+        Logger.autoconsent.debug("Recording popup handled: CMP \(cmpName) on \(String(describing: self.topUrl))")
+        lastHandledCMPName = cmpName
+    }
+
+    /// Clears the reload loop detection state
+    private func clearReloadLoopState() {
+        lastHandledCMPName = ""
+        reloadLoopDetected = false
+    }
+
+    /// Compares two URL strings ignoring query parameters and fragments
+    /// - Parameters:
+    ///   - url1: First URL string
+    ///   - url2: Second URL string
+    /// - Returns: True if protocol, host, and path match, false otherwise
+    private func urlsMatchIgnoringQuery(_ url1: URL?, _ url2: URL?) -> Bool {
```

Comment:
these sorts of things usually belong as an extension on URL

I see we already have trimmingQueryItemsAndFragment() from URL+Trimming.swift

it’s in the PrivacyDashboard directory, though, so should move it